R/decostand.R

```diff
@@ -260,3 +277,348 @@
+# }
+# @references
+# Raghunandan H. Keshavan, Andrea Montanari, Sewoong Oh (2010).
+# Matrix Completion From a Few Entries.
+# IEEE Transactions on Information Theory 56(6):2980--2998.
+#
+.OptSpace <- function(A, ropt=NA, niter=50, tol=1e-6, showprogress=FALSE){
+  ## Preprocessing : A     : partially revelaed matrix
+  if (!is.matrix(A)){
+    stop("* .OptSpace : an input A should be a matrix.")
+  }
+  if (any(is.infinite(A))){
+    stop("* .OptSpace : no infinite value in A is allowed.")
+  }
+  if (!any(is.na(A))){
+    stop("* .OptSpace : there is no unobserved values as NA.")
+  }
+  idxna = (is.na(A))
+  M_E = array(0,c(nrow(A),ncol(A)))
+  M_E[!idxna] = A[!idxna]
+  
+  ## Preprocessing : size information
+  n = nrow(A)
+  m = ncol(A)
+  
+  ## Preprocessing : other sparse-related concepts
+  nnZ.E = sum(!idxna)
+  E = array(0,c(nrow(A),ncol(A))); E[!idxna] = 1
+  eps = nnZ.E/sqrt(m*n)
+  
+  ## Preprocessing : ropt  : implied rank
+  if (is.na(ropt)){
+    if (showprogress){
+      message("* .OptSpace: Guessing an implicit rank.")
+    }
+    r = min(max(round(.guess_rank(M_E,nnZ.E)), 2), m-1)
+    if (showprogress){
+      message(paste0('* .OptSpace: Guessing an implicit rank: Estimated rank : ',r))
+    }
+  } else {
+    r = round(ropt)
+    if ((!is.numeric(r))||(r<1)||(r>m)||(r>n)){
+      stop("* .OptSpace: ropt should be an integer in [1,min(nrow(A),ncol(A))].")
+    }
+  }
+  ## Preprocessing : niter : maximum number of iterations
+  if ((is.infinite(niter))||(niter<=1)||(!is.numeric(niter))){
+    stop("* .OptSpace: invalid niter number.")
+  }
+  niter = round(niter)
+  
+  m0 = 10000
+  rho = 0
+  
+  ## Main Computation
+  rescal_param = sqrt(nnZ.E*r/(norm(M_E,'f')^2))
+  M_E = M_E*rescal_param
+  
+  # 1. Trimming
+  if (showprogress){
+    message("* .OptSpace: Step 1: Trimming ...")
+  }
+  M_Et = M_E
+  d  = colSums(E)
+  d_ = mean(d)
+  for (col in 1:m){
+    if (sum(E[,col])>(2*d_)){
+      listed = which(E[,col]>0)
+      p = sample(1:length(listed),length(listed))
+      M_Et[listed[p[ceiling(2*d_)]]:n,col] = 0
+    }
+  }
+  
+  d  = rowSums(E)
+  d_ = mean(d)
+  for (row in 1:n){
+    if (sum(E[row,])>2*d_){
+      listed = which(E[row,]>0)
+      p = sample(1:length(listed),length(listed))
+      M_Et[row,listed[p[ceiling(2*d_)]]:m] = 0
+    }
+  }
+  
+  # 2. SVD
+  if (showprogress){
+    message("* .OptSpace: Step 2: SVD ...")
+  }
+  svdEt = svd(M_Et)
+  X0 = svdEt$u[,1:r]
+  S0 = diag(svdEt$d[1:r])
+  Y0 = svdEt$v[,1:r]
+  
+  # 3. Initial Guess
+  if (showprogress){
+    message("* .OptSpace: Step 3: Initial Guess ...")
+  }
+  X0 = X0*sqrt(n)
+  Y0 = Y0*sqrt(m)
+  S0 = S0/eps
+  
+  # 4. Gradient Descent
+  if (showprogress){
+    message("* .OptSpace: Step 4: Gradient Descent ...")
+  }
+  X = X0
+  Y = Y0
+  S = .aux_getoptS(X,Y,M_E,E)
+  
+  # initialize
+  dist = array(0,c(1,(niter+1)))
+  dist[1] = norm((M_E - (X%*%S%*%t(Y)))*E,'f')/sqrt(nnZ.E)
```

Comment:
Formatting again: use spaces around %*% for readability (still quite OK here, but becomes awkward when you have list elements with $-signs.

Reply:
true, added